### ISSUE TITLE
[Bug] Custom Service Account being overwritten by operator

### ISSUE THREAD

**Issue by Fryuni** [strimzi/strimzi-kafka-operator#3300]
**Describe the bug**
Before 0.18.0 if the KafkaConnect service account already existed it would just use it, now it is overwriting it with the strimzi-managed metadata.

**To Reproduce**
1. Create a ServiceAccount `some-name-connect` with any custom annotation and/or labels
2. Create a KafkaConnect `some-name`, any configuration works
3. The ServiceAccount metadata will be replaced

**Expected behavior**
The ServiceAccount should be left unchanged like before.

**Environment (please complete the following information):**
 - Strimzi version: 0.18.0
 - Installation method: Helm Chart
 - Kubernetes cluster: v1.14.10-gke.36
 - Infrastructure: GKE

**Additional context**
We use custom annotations on the service account to allow identity sharing from the cluster to the external resources (external to the cluster) using [GKE's Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity)

**Comment by scholzj** [comment]
I don't think this is a bug. I think this is expected. But if you want we can have a look at it and add a possibility to specify annotations for service account in the `template` sections (see https://strimzi.io/docs/operators/latest/full/using.html#assembly-customizing-deployments-str). Would that work for you?

**Comment by barrydobson** [comment]
I'm also seeing this problem where we need to attach annotations to the Service Account to allow IAM authentication. 
In my case, service accounts which are annotated this way are created and managed by EKSCTL command on cluster deployment, so being able to add an annotation to template wouldn't work for us.
Ideally, if the service account exists it should be used as is without dropping any labels/annotations

**Comment by Fryuni** [comment]
Yeah, something like this would work:
```
...
spec:
  template:
    serviceAccount:
      metadata: # Maybe this level is not needed for a ServiceAccount resource
        labels: {}
        annotations: {}
```
The labels from the KafkaConnect resource are passed to the ServiceAccount, but it would be better to add labels there if they are not required on the other resources.

If this is the expected behavior now, I think it should be noted as a breaking change and added to the [upgrade instructions](https://strimzi.io/docs/operators/0.18.0/deploying.html#assembly-upgrade-resources-str) as it does indeed break production deployments.

We worked around that by setting a post-intall/post-upgrade script to edit the resource after it is created/edited by strimzi adding the annotation

**Comment by scholzj** [comment]
> If this is the expected behavior now, I think it should be noted as a breaking change and added to the upgrade instructions as it does break production deployments.

Not, this is not supported right now. All I'm saying is that we can look at adding this if it helps with your issue on GKE.

> If this is the expected behavior now, I think it should be noted as a breaking change and added to the upgrade instructions as it does break production deployments.

I do not think there is anything breaking. This works the same as it worked before as far as Strimzi is involved. If this worked differently before the only reason for it could be change in the Fabric8 Kubernetes client or in Kubernetes it self. 

> We worked around that by setting a post-intall/post-upgrade script to edit the resource after it is created/edited by strimzi adding the annotation

I'm a bit surprised this works. Every loop of the operator does the same operation (basically applies our desired version of given resource). So it should either override this every 2 minutes or not touch it at all.

**Comment by Fryuni** [comment]
> I do not think there is anything breaking. This works the same as it worked before as far as Strimzi is involved. If this worked differently before the only reason for it could be change in the Fabric8 Kubernetes client or in Kubernetes it self.

The behavior did change in a way that breaks deployments, from just using resources that were already created to replacing it upon creation of the KafkaConnect resource. I'm not saying it is caused by some change _in strimzi itself_, it might be due to a change on a dependency, but wherever the change happened it is breaking use cases for strimzi and should be noted as such.

> I'm a bit surprised this works. Every loop of the operator does the same operation (basically applies our desired version of given resource). So it should either override this every 2 minutes or not touch it at all.

It is only affecting the ServiceAccount when the KafkaConnect resource changes, apparently it is not applied on the loop. The deployment is, any manual change is undone on the next reconciliation.

**Comment by scholzj** [comment]
> The behavior did change in a way that breaks deployments, from just using resources that were already created to replacing it upon creation of the KafkaConnect resource. I'm not saying it is caused by some change in strimzi itself, it might be due to a change on a dependency, but wherever the change happened it is breaking use cases for strimzi and should be noted as such.

Do you have any logs for that? Which version worked for you?

> It is only affecting the ServiceAccount when the KafkaConnect resource changes, apparently it is not applied on the loop. The deployment is, any manual change is undone on the next reconciliation.

You mean that it happens only after change to KafkaConnect CR and not during the periodical reconciles?

**Comment by Fryuni** [comment]
> Do you have any logs for that? Which version worked for you?

It worked on 0.17.0. I can get the logs from the time it was running, but I looked over them when I was trying to understand this, there is nothing regarding the service accounts there, probably only on debug level, but I never used debug level when it was working (logically). I can try to find some time to set a new k8s cluster with version 0.17.0 and 0.18 with debug logs.

> You mean that it happens only after change to KafkaConnect CR and not during the periodical reconciles?

Yes, the operator only changes the ServiceAccount when I change the labels on the KafkaConnect resource, which is then replicated to the ServiceAccount and therefore overwrite the custom annotations. If I add a custom label/annotation on the Deployment it is replaced on the next reconciliation, but my custom annotation on the ServiceAccount is ignored

**Comment by scholzj** [comment]
@Fryuni This got to the top of my todo list now and I was looking at it.  But I cannot reproduce it with any version - neither 0.18.0 not master. The code which is handling the service account is explicitly not patching them because it causes changes to the service account tokens: https://github.com/strimzi/strimzi-kafka-operator/blob/19274ab63c9baf9ecfce1c993a70634758af8f6d/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ServiceAccountOperator.java#L35 So for me if the SA exists when I create the Kafka resource, it gets never changed, it doesn't change, it never adds any labels and it doesn't set the owner reference etc. 

Can you double check the problem and eventually provide the debug logs?

**Comment by Fryuni** [comment]
It just happened on our deployment this Monday, we changed the memory limit on the KafkaConnect resource and the annotation disappeared from the SA. We'll have another change this week so when I get to this task I'll do it with debug logs on our test environment and post it here.

Strangely, id didn't happen on our staging environment for the exact same deploy and the same diff.

**Comment by scholzj** [comment]
Well, let's see in the DEBUG log file.

**Comment by neemiasjnr** [comment]
I have hit this issue today, as mentioned before, adding `serviceAccount` to the `KafkaConnectTemplate` would work. I can try to raise a PR if you agree with that change @scholzj

**Comment by scholzj** [comment]
Feel free to have a look at it. PRs are always welcome. Just keep in mind that the ServiceAccount does not reconcile as other resources because that regenerates the tokens which causes problems: https://github.com/strimzi/strimzi-kafka-operator/blob/675057e50035bbae4d7b73da1b8149963358bbec/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ServiceAccountOperator.java#L33. So if you just add it to the template, it would create it with any options but not update. I'm not sure if there is some easy way how to solve it ... maybe we would need to diff it to see if something really changed. But overall I do not think there is any problem with having the template section for service accounts.

**Comment by scholzj** [comment]
(And feel free to let me know here or on Slack if you need any help with the PR)
